six/modules/c++/six.sicd/source/AreaPlaneUtility.cpp

```diff
@@ -0,0 +1,260 @@
+}
+
+types::RgAz<std::vector<double> > AreaPlaneUtility::computeMetersFromCenter(
+        const ComplexData& data,
+        const types::RowCol<Vector3>& unitVectors)
+{
+    std::vector<Vector3> inPlaneCornersECEF =
+            computeInPlaneCorners(data, unitVectors);
+
+    types::RgAz<std::vector<double> > metersFromCenter(
+            std::vector<double>(inPlaneCornersECEF.size()),
+            std::vector<double>(inPlaneCornersECEF.size()));
+    for (size_t ii = 0; ii < inPlaneCornersECEF.size(); ++ii)
+    {
+        metersFromCenter.rg[ii] = unitVectors.row.dot(
+                inPlaneCornersECEF[ii] - data.geoData->scp.ecf);
+        metersFromCenter.az[ii] = unitVectors.col.dot(
+                inPlaneCornersECEF[ii] - data.geoData->scp.ecf);
+    }
+
+    std::sort(metersFromCenter.rg.begin(), metersFromCenter.rg.end());
+    std::sort(metersFromCenter.az.begin(), metersFromCenter.az.end());
+
+    return metersFromCenter;
+}
+
+types::RowCol<size_t> AreaPlaneUtility::derivePlaneDimensions(
+        const types::RgAz<std::vector<double> >& sortedMetersFromCenter,
+        const RowColDouble& spacing)
+{
+    const types::RowCol<size_t> lastIndex(
+            sortedMetersFromCenter.rg.size() - 1,
+            sortedMetersFromCenter.az.size() - 1);
+    const double rowDistance = std::abs(
+            sortedMetersFromCenter.rg[lastIndex.row] -
+            sortedMetersFromCenter.rg[0]);
+    const double colDistance = std::abs(
+            sortedMetersFromCenter.az[lastIndex.col] -
+            sortedMetersFromCenter.az[0]);
+
+    types::RowCol<size_t> dimensions;
+    dimensions.row = static_cast<size_t>(
+            std::ceil(std::max<double>(1.0, rowDistance / spacing.row)));
+    dimensions.col = static_cast<size_t>(
+            std::ceil(std::max<double>(1.0, colDistance / spacing.col)));
+    return dimensions;
+}
+
+RowColDouble AreaPlaneUtility::deriveReferencePoint(
+        const types::RgAz<std::vector<double> >& sortedMetersFromCenter,
+        const RowColDouble& spacing,
+        const AreaPlane& areaPlane)
+{
+    RowColDouble referencePoint;
+    referencePoint.row = std::abs(sortedMetersFromCenter.rg[0]) / spacing.row
+                    + areaPlane.xDirection->first;
+    referencePoint.col = std::abs(sortedMetersFromCenter.az[0]) / spacing.col
+                    + areaPlane.yDirection->first;
+    return referencePoint;
+}
+
+void AreaPlaneUtility::setAreaPlane(ComplexData& data,
+        bool includeSegmentList)
+{
+    if (!data.radarCollection.get())
+    {
+        data.radarCollection.reset(new RadarCollection());
+    }
+    if (!data.radarCollection->area.get())
+    {
+        data.radarCollection->area.reset(new Area());
+    }
+    if (!data.radarCollection->area->plane.get())
+    {
+        data.radarCollection->area->plane.reset(new AreaPlane());
+        AreaPlane& areaPlane = *data.radarCollection->area->plane;
+        deriveAreaPlane(data, areaPlane);
+        if (includeSegmentList)
```

Comment:
Ahh I see.  Yeah in `deriveAreaPlane()` just set `areaPlane.segmentList` and then in `setAreaPlane()` set `imageFormation.segmentIdentifier` (in all cases, only set that stuff if `includeSegmentList` is set).